rfcs/2020-04-15-wasm-foreign-modules.md

```diff
@@ -0,0 +1,559 @@
+* You want to write a complex transform in your favorite language.
+
+Here's some examples of where WASM probably isn't right for you:
+
+* You only need basic functionality already offered by Vector.
+* Your desired language doesn't support WASM as a compile target.
+* You are using Vector on a non-Linux X86-64 bit compatible target. (Lucet, our Engine, only supports Linux at this
+current time. Work on other operating systems is ongoing!)
+* You need a battle-proven system, WASM support in Vector is young!
+
+
+### How Vector modules work
+
+Vector modules are `.wasm` files (WASI or not) exposing a predefined interface. We provide a `foreign_modules::guest`
+module for Rust projects located in `lib/foreign-modules` of the source tree, and we will be providing bindings for
+other languages as we see user demand.
+
+Modules are sandboxed and can only interact with Vector over the Foreign Function Interface (FFI) which Vector exposes.
+For each instance of a module, Vector holds some context. For example, when a transform module processes an event, the
+context of the module contains an `Event` type. The guest can use calls like `get("foo")` and Vector perform the action
+and return the result to the guest.
+
+A guest module is not able to access the memory of the Vector instance, but Vector is free to modify the guest memory.
+For example, behind the scenes of a `get` call, the Guest must allocate some memory for Vector to write the result into.
+
+Due to the current semantics of the `vector::Event` type, passing data between WASM modules and the Vector host involves
+serialization to JSON. We're investigating ways to speed this up.
+
+* It may be possible to add a `repr(C)` flag to Event.
+* It may be worthwhile to explore an `Event` refactor.
+* Other serialization options may be much faster but less compatible.
+
+
+### Initialization
+
+
+Regardless of whether of not the module uses WASI, or has access to a language specific binding, it can interact with
+Vector. Vector works by invoking the modules public interface. First, as soon as the module is loaded, `register` is
+called once, globally, per module. In this call, the module can configure how Vector sees it and talks to it. During
+this time, the module can do any one time setup it needs. (Eg. Opening a socket, a file).
+
+If using the Vector guest API this can be done via:
+
+```rust
+use foreign_modules::{Registration, roles::Sink};
+
+#[no_mangle]
+pub extern "C" fn init() -> *mut Registration {
+ &mut Registration::transform()
+    .set_wasi(true) as *mut Registration
+}
+```
+
+What happens next depends on which type of module it is.
+
+### Module types
+
+* `Transform`: Modules of this role can be used as a transform `type`.
+
+  ```rust
+  use foreign_modules::hostcall;
+
+  // TODO: Add better FFI error handling!
+  #[no_mangle]
+  pub extern "C" fn process() -> usize {
+      if let Some(value) = hostcall::get("field").unwrap() {
```

Comment:
Should manipulations on events, such as dropping events or emitting multiple events out of a single one, be supported by transforms too?

Reply:
I'd like to find API parity with the LUA transform, so yes. I think this can be reworked a bit. If we can make `Event` a `repr(C)` type that would help a lot. Let me experiment with this a bit?

Reply:
Agreed! We can definitely support those through hostcalls, it's just a matter of figuring out an ergonomic API that works!

Reply:
Can we use the new lua changes as inspiration? I feel like the `emit` function used there works well.

Reply:
Yeah that's basically what I want to copy. :)

Reply:
It works!